src/transaction/log_append.cpp

```diff
@@ -0,0 +1,1857 @@
+{
+#if defined (SERVER_MODE)
+  if (thread_p == NULL)
+    {
+      thread_p = thread_get_thread_entry_info ();
+    }
+
+  if (thread_p == NULL)
+    {
+      return NULL;
+    }
+  else
+    {
+      if (thread_p->log_zip_redo == NULL)
+	{
+	  thread_p->log_zip_redo = log_zip_alloc (IO_PAGESIZE, true);
+	}
+      return (LOG_ZIP *) thread_p->log_zip_redo;
+    }
+#else
+  return log_zip_redo;
+#endif
+}
+
+/*
+ * log_append_realloc_data_ptr -
+ *
+ * return:
+ *
+ *   data_length(in):
+ *   length(in):
+ *
+ * NOTE:
+ */
+static bool
+log_append_realloc_data_ptr (THREAD_ENTRY *thread_p, int length)
+{
+  char *data_ptr;
+  int alloc_len;
+#if defined (SERVER_MODE)
+  if (thread_p == NULL)
+    {
+      thread_p = thread_get_thread_entry_info ();
+    }
+
+  if (thread_p == NULL)
+    {
+      return false;
+    }
+
+  if (thread_p->log_data_length < length)
+    {
+      alloc_len = ((int) CEIL_PTVDIV (length, IO_PAGESIZE)) * IO_PAGESIZE;
+
+      data_ptr = (char *) realloc (thread_p->log_data_ptr, alloc_len);
+      if (data_ptr == NULL)
+	{
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, (size_t) alloc_len);
+	  if (thread_p->log_data_ptr)
+	    {
+	      free_and_init (thread_p->log_data_ptr);
+	    }
+	  thread_p->log_data_length = 0;
+	  return false;
+	}
+      else
+	{
+	  thread_p->log_data_ptr = data_ptr;
+	  thread_p->log_data_length = alloc_len;
+	}
+    }
+  return true;
+#else
+  if (log_data_length < length)
+    {
+      alloc_len = ((int) CEIL_PTVDIV (length, IO_PAGESIZE)) * IO_PAGESIZE;
+
+      data_ptr = (char *) realloc (log_data_ptr, alloc_len);
+      if (data_ptr == NULL)
+	{
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, (size_t) alloc_len);
+	  if (log_data_ptr)
+	    {
+	      free_and_init (log_data_ptr);
+	    }
+	  log_data_length = 0;
+	  return false;
+	}
+      else
+	{
+	  log_data_ptr = data_ptr;
+	  log_data_length = alloc_len;
+	}
+    }
+
+  return true;
+#endif
+}
+
+/*
+ *
+ */
+static char *
+log_append_get_data_ptr (THREAD_ENTRY *thread_p)
+{
+#if defined (SERVER_MODE)
+  if (thread_p == NULL)
+    {
+      thread_p = thread_get_thread_entry_info ();
+    }
+
+  if (thread_p == NULL)
+    {
+      return NULL;
+    }
+  else
+    {
+      if (thread_p->log_data_ptr == NULL)
+	{
+	  thread_p->log_data_length = IO_PAGESIZE * 2;
+	  thread_p->log_data_ptr = (char *) malloc (thread_p->log_data_length);
+
+	  if (thread_p->log_data_ptr == NULL)
+	    {
+	      thread_p->log_data_length = 0;
+	      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1,
+		      (size_t) thread_p->log_data_length);
+	    }
+	}
+      return thread_p->log_data_ptr;
+    }
+#else
+  return log_data_ptr;
+#endif
+}
+
+static void
+log_prior_lsa_append_align ()
+{
+  log_Gl.prior_info.prior_lsa.offset = DB_ALIGN (log_Gl.prior_info.prior_lsa.offset, DOUBLE_ALIGNMENT);
+  if ((size_t) log_Gl.prior_info.prior_lsa.offset >= (size_t) LOGAREA_SIZE)
```

Comment:
Old version (macro) use compare as "int";
I think it is safer the old way ; what if  offset is negative ?

Reply:
If offset is negative, our troubles are bigger than conversions to size_t. Added safe-guard to always check offset is 0 or positive. Do you think it is enough?

Reply:
Negative offset would mean uninitialized prior_lsa.
I guess this is enough for this case.
But, generally speaking I would prefer "int" compare, since it is less probable to have very large LOGAREA_SIZE (is detected a lot faster) than a negative value in offest of a LSA.